### PR TITLE
fix: use yarn instead of npx with yarn.lock

### PR DESCRIPTION
**Pull request by MikeMcC399**
This PR resolves an issue calling commands `cypress cache list` and `cypress cache verify` if the project is using [Yarn Plug'n'Play](https://yarnpkg.com/features/pnp) (introduced through PR https://github.com/cypress-io/github-action/pull/599 from @PilotConway). In this case the commands were not finding the installed version of Cypress.

## Changes

The following commands, which are called internally by the action, are replaced if the action is working from a `yarn.lock` file. This is the case for [Yarn Classic](https://classic.yarnpkg.com/) and [Yarn Modern](https://yarnpkg.com/) projects. Although the issue only affects [Yarn Modern](https://yarnpkg.com/) using [Yarn Plug'n'Play](https://yarnpkg.com/features/pnp) configuration, the change is activated for **all** types of Yarn projects.

The following commands are retained for `npm` and `pnpm` projects:

```bash
npx cypress cache list
npx cypress verify
```

For Yarn projects they are changed to use instead:

```bash
yarn cypress cache list
yarn cypress verify
```

## Issue

When [.github/workflows/example-yarn-modern-pnp.yml](https://github.com/cypress-io/github-action/blob/master/.github/workflows/example-yarn-modern-pnp.yml) is run, it shows `npx` being called and not finding the installed version of cypress.

For example job [4959218510](https://github.com/cypress-io/github-action/actions/runs/4959218510) shows:

```text
/usr/local/bin/npx cypress cache list
npm WARN exec The following package was not found and will be installed: cypress@12.12.0
┌─────────┬───────────────────┐
│ version │ last used         │
├─────────┼───────────────────┤
│ 12.11.0 │ a few seconds ago │
├─────────┼───────────────────┤
│ 12.12.0 │ 3 days ago        │
└─────────┴───────────────────┘
/usr/local/bin/npx cypress verify
```

With [Yarn Plug'n'Play](https://yarnpkg.com/features/pnp) there is no directory `node_modules` set up and therefore [npx](https://docs.npmjs.com/cli/v9/commands/npx) cannot find `cypress` and it installs instead the latest version, which may or may not correspond to the version specified in `yarn.lock`.

## Verification

Check the workflow results:

- [example-basic.yml](https://github.com/cypress-io/github-action/blob/master/.github/workflows/example-basic.yml)
- [example-basic-pnpm.yml](https://github.com/cypress-io/github-action/blob/master/.github/workflows/example-basic-pnpm.yml)
- [example-yarn-classic.yml](https://github.com/cypress-io/github-action/actions/workflows/example-yarn-classic.yml)
- [example-yarn-modern.yml](https://github.com/cypress-io/github-action/actions/workflows/example-yarn-modern.yml)
- [example-yarn-modern-pnp.yml](https://github.com/cypress-io/github-action/actions/workflows/example-yarn-modern-pnp.yml)

1. All workflows should be successful.
2. [example-basic.yml](https://github.com/cypress-io/github-action/blob/master/.github/workflows/example-basic.yml) and [example-basic-pnpm.yml](https://github.com/cypress-io/github-action/blob/master/.github/workflows/example-basic-pnpm.yml) should use `npx` when calling  `cypress cache list` or `cypress verify`.
3. Each of the Yarn workflows should use `yarn` when calling  `cypress cache list` and `cypress verify`. Only one version of Cypress should be shown by `cypress cache list`.

### CODE REVIEW

**Comment by MikeMcC399** [comment]
Job [8937982597](https://github.com/cypress-io/github-action/actions/runs/4991464444/jobs/8937982597) shows successful selection of `yarn cypress cache list` and `yarn cypress verify` from [example-yarn-modern-pnp.yml](https://github.com/cypress-io/github-action/actions/workflows/example-yarn-modern-pnp.yml). Only one version of Cypress is show in the cache list, which is correct.

```text
/usr/local/bin/yarn cypress cache list
┌─────────┬───────────────────┐
│ version │ last used         │
├─────────┼───────────────────┤
│ 12.11.0 │ a few seconds ago │
└─────────┴───────────────────┘
/usr/local/bin/yarn cypress verify

[STARTED] Task without title.
[SUCCESS] Task without title.

```

**Comment by PilotConway** [comment]
Question: Any plans to use this same approach on the cypress test run itself? That seems to be the only other place hardcoded to `npx`, and fixing that line would remove the requirement for my `command` fix completely. 

https://github.com/MikeMcC399/github-action/blob/2a93c9f278747ca1ce07982dbafb7f50f6823777/index.js#L634-L647

I did run this on our internal code to test and it was able to find and use my installed cypress version so this PR is working great for us! Thanks!

**Comment by PilotConway** [comment]
@MikeMcC399 Awesome, sounds good! Thanks for putting in fixes for this stuff, I know my team definitely appreciates it. 😄

**Comment by MikeMcC399** [comment]
@nagash77 
This PR from mid May could also benefit from some attention. So far it has only been assigned, then re-assigned.

**Comment by MikeMcC399** [comment]
Hi @warrensplayer 

Is there any chance you could review this PR sometime soon?

**Comment by warrensplayer** [comment]
@MikeMcC399 Looking at this now.  Sorry for the delay.

File: dist/index.js [hunk 1]
```diff
@@ -74457,6 +74457,11 @@ const useYarn = () => fs.existsSync(yarnFilename)
 const usePnpm = () => fs.existsSync(pnpmLockFilename)
 const useNpm = () => fs.existsSync(packageLockFilename)
 
+let runPrefix = 'npx'
+if (useYarn()) {
+  runPrefix = 'yarn'
+}
```
Comment (warrensplayer) [inline dist/index.js:74463]:
```suggestion
const runPrefix = useYarn() ? 'yarn' : 'npx'
```

How about this to use a `const` versus `let` variable?

File: index.js [hunk 2]
```diff
@@ -106,6 +106,11 @@ const useYarn = () => fs.existsSync(yarnFilename)
 const usePnpm = () => fs.existsSync(pnpmLockFilename)
 const useNpm = () => fs.existsSync(packageLockFilename)
 
+let runPrefix = 'npx'
+if (useYarn()) {
+  runPrefix = 'yarn'
+}
+
```
Comment (warrensplayer) [inline index.js:113]:
```suggestion
const runPrefix = useYarn() ? 'yarn' : 'npx'
```

Same suggestion here.

**Review by lmiller1990 — APPROVED** [review]

File: index.js (hunk 2 above)
Comment (MikeMcC399) [reply index.js]: @warrensplayer 
Many thanks for your code simplification suggestion! I have committed and tested this.

File: dist/index.js (hunk 1 above)
Comment (MikeMcC399) [reply dist/index.js]: @warrensplayer 

You have commented here on the compiled code, so effectively it's a duplicate of your other comment.

The change you requested has been implemented by editing the source code `index.js` and then by executing `npm run build` which updates `dist/index.js`,

**Comment by MikeMcC399** [comment]
@warrensplayer

> Looking at this now. Sorry for the delay.

Thank you for your review! I included your suggestion for code optimization. All tests are passing, including the specific verification tests I listed in the original post.

**Review by AtofStryker — APPROVED** [review]

**Review by warrensplayer — APPROVED** [review]

**Comment by MikeMcC399** [comment]
- Instead of re-applying this PR at a later date as a breaking change, I suggest to implement full support for Yarn Modern, making no changes to the way Yarn Classic projects are handled. See feature request https://github.com/cypress-io/github-action/issues/969.

**Pull request merged**